src/rust/sqs-executor/src/redis_cache.rs

```diff
@@ -136,53 +230,54 @@ impl RedisCache {
-                    continue;
-                }
-                lru_cache.insert(identity_bytes.clone(), ());
-                identities_to_check.push(identity_bytes);
-            }
-        }
+
+            // for each of the identities:
+            //      1. update the lru cache status
+            //      2. if key is new (is_none()), collect into a Vec
+            identities
+                .into_iter()
+                .filter(|identity| lru_cache.put(identity.clone(), ()).is_none())
+                .collect()
+        };
+
         if identities_to_check.is_empty() {
             return Ok(());
         }
-        let identities = identities_to_check;
 
         let client_pool = self.connection_pool.clone();
         let task_start = Instant::now();
 
-        tokio::time::timeout(
-            Duration::from_millis(300),
-            tokio::spawn(async move {
-                let mut builder = MSetBuilder::new();
-                for identity in identities.iter() {
-                    builder = builder.set(identity, b"1");
-                }
-                let mut client = client_pool.get().await;
-                let res = client.mset(builder).await;
-                match res {
-                    Err(e) => {
-                        let elapsed = task_start.elapsed();
-                        if elapsed >= Duration::from_millis(300) {
-                            let elapsed_ms = elapsed.as_millis() as u64;
-                            error!(
-                                error = e.to_string().as_str(),
-                                elapsed_ms = elapsed_ms,
-                                "redis mset failed outside of ttl"
-                            );
-                            Ok(())
-                        } else {
-                            Err(e)
-                        }
+        tokio::spawn(async move {
+            let mut client = client_pool.get().await;
+            let mut mset_builder = MSetBuilder::new();
+
+            for identity in &identities_to_check {
+                mset_builder = mset_builder.set(identity, b"1");
+            }
+
+            match client.mset(mset_builder).await {
+                Ok(_) => Ok(()),
+                Err(e) => {
+                    let elapsed = task_start.elapsed();
+
+                    if elapsed >= Duration::from_millis(300) {
+                        let elapsed_ms = elapsed.as_millis() as u64;
+                        error!(
+                            error = e.to_string().as_str(),
+                            elapsed_ms = elapsed_ms,
+                            "redis mset failed outside of ttl"
+                        );
+                        Ok(())
+                    } else {
+                        Err(e)
                     }
-                    Ok(_) => Ok(()),
                 }
-            }),
-        )
+            }
+        })
+        .timeout(Duration::from_millis(300))
         .await???;
```

Comment:
I would really be curious to see if there's a way to reformulate this so that three `?` aren't required. That is so surprising to see that it strikes me immediately as a code smell.

It may very well be valid, but if so, it's the kind of thing I'd like to see a comment for:

```rust
// I know what you're thinking... THREE question mark operators? 
//
// Well, gentle reader, let me tell you a story...
```